openscm/adapters/dice.py

```diff
@@ -0,0 +1,388 @@
+    # Initial lower stratum temperature change (rel. to 1900)
+    "T_ocean0": (0.0068, "degC"),
+    # Lower bound for lower stratum temperature change (rel. to 1900)
+    "T_ocean_lower": (-1, "degC"),
+    # Upper bound for lower stratum temperature change (rel. to 1900)
+    "T_ocean_upper": (20, "degC"),
+    # Carbon cycle transition matrix
+    "b12": (0.0181, ""),
+    # Carbon cycle transition matrix
+    "b23": (0.00071, ""),
+    # Climate equation coefficient for upper level
+    "c1": (0.0222, "degC*m^2/W"),
+    # Transfer coefficient upper to lower stratum
+    "c3": (0.09175, "W/m^2/degC"),
+    # Transfer coefficient for lower level
+    "c4": (0.00487, ""),
+    # Forcings of equilibrium CO2 doubling
+    "fco22x": (3.8, "W/m^2"),
+    # 2010 forcings of non-CO2 GHG
+    "fex0": (0.25, "W/m^2"),
+    # 2100 forcings of non-CO2 GHG
+    "fex1": (0.7, "W/m^2"),
+    # Equilibrium climate sensitivity
+    "t2xco2": (2.9, "degC"),
+}
+
+
+class DICE(Adapter):
+    """
+    Adapter for the climate component from the Dynamic Integrated Climate-Economy (DICE)
+    model.
+
+    TODO recalculate increase to absolute values
+    TODO still calibrated to initial year 2010
+    TODO What about timeseries parameters when run times change?
+    """
+
+    _period_length: int
+    """Period length in seconds (currently 1yr)"""
+
+    _timestep: int
+    """Current time step"""
+
+    _timestep_count: int
+    """Total number of time steps"""
+
+    _values: Any
+    """Parameter values"""
+
+    _views: Any
+    """Parameter views"""
+
+    _year2100: int
+    """Timestamp of 2100-01-01"""
+
+    def _initialize_model(self) -> None:
+        """
+        Initialize the model.
+        """
+        self._period_length = YEAR
+        self._year2100 = int(datetime(2100, 1, 1).timestamp())  # TODO 365 days-year
+
+        parameter_names = list(MODEL_PARAMETER_DEFAULTS.keys()) + [
+            "E",
+            "M_atm",
+            "M_l",
+            "M_u",
+            "T_atm",
+            "T_ocean",
+            "force",
+            "b11",
+            "b21",
+            "b22",
+            "b32",
+            "b33",
+        ]
+        self._values = namedtuple("DICEValues", parameter_names)
+        self._views = namedtuple("DICEViews", parameter_names)
+
+        for name, (default, unit) in MODEL_PARAMETER_DEFAULTS.items():
+            setattr(self._values, name, default)
+            self._parameters.get_writable_scalar_view(
+                ("DICE", name), ("World",), unit
+            ).set(default)
+            setattr(
+                self._views,
+                name,
+                self._parameters.get_scalar_view(("DICE", name), ("World",), unit),
+            )
+
+    def _initialize_model_input(self) -> None:
+        pass
+
+    def _initialize_run_parameters(self) -> None:
+        self._timestep = 0
+        self._timestep_count = (
+            self._stop_time - self._start_time
+        ) // self._period_length
```

Comment:
maybe +1 here so it runs over the stop_time rather than stopping before it (it's easy to filter out an extra timestep after if you really don't want it but running again can be annnoying)?

Reply:
Oh, good point. Sorry, missed that we said including the stop time. Will also add a base test for that.

Reply:
(keeping open as areminder to myself)